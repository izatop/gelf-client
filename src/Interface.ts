import {Level} from "./Level";
import {TransportCtor} from "./TransportAbstract";

export type ConnectionOptions = Readonly<{
    host: string;
    port: number;
    protocol: string;
    compress: boolean;
    minCompressSize: number;
    maxChunkSize: number;
    strictChecks: boolean;
}>;

export interface IMessageRest {
    [key: string]: any;
}

export interface IMessage {
    app?: string;
    level?: Level;
    message: string;
    description?: string;
    file?: string;
    line?: number;
}

export type MessageDefaults = Partial<IMessage> & IMessageRest;

export interface IEnvelope {
    level: Level;
    version: string;
    timestamp: number;
    short_message: string;
    full_message?: string;
    host?: string;
    file?: string;
    line?: number;
    [key: string]: any;
}
