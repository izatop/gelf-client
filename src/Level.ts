export const enum Level {
    EMERGENCY,
    ALERT,
    CRITICAL,
    ERROR,
    WARNING,
    NOTICE,
    INFO,
    DEBUG,
}
