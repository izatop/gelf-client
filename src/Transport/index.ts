export * from "./TCPTransport";
export * from "./UDPTransport";
