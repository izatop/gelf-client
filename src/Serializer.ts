import {randomBytes} from "crypto";
import {compress, GELF_MAGIC_NO} from "./config";
import {ConnectionOptions} from "./Interface";

export class Serializer {
    public readonly options: ConnectionOptions;

    constructor(options: ConnectionOptions) {
        this.options = options;
    }

    public async serialize(data: Buffer) {
        if (data.length <= this.options.maxChunkSize) {
            return [data];
        }

        if (!this.options.compress && data.length >= this.options.minCompressSize) {
            return this.createChunks(data);
        }

        return this.createChunks(await compress(data));
    }

    protected createChunks(data: Buffer) {
        const bufferSize = this.options.maxChunkSize - 12;
        const chunksCount = Math.ceil(data.length / bufferSize);
        if (chunksCount > 128) {
            throw new Error(`Cannot log messages bigger than ${bufferSize * 128} bytes`);
        }

        const id = randomBytes(8);
        const chunks: Buffer[] = [];
        const length = data.length;

        let sequence = 0;
        while (sequence < chunksCount) {
            const offset = sequence * bufferSize;
            const readBytes = offset + bufferSize < length
                ? offset + bufferSize
                : length;

            const chunk = Buffer.from([
                ...GELF_MAGIC_NO, // GELF Magic bytes
                ...id, // Message identifier
                sequence++, // Sequence
                chunksCount, // Number of chunks
                ...data.slice(offset, readBytes), // Message chunk
            ]);

            chunks.push(chunk);
        }

        return chunks;
    }
}
