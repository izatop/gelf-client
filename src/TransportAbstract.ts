import {EventEmitter} from "events";
import {Url} from "url";
import {ConnectionOptions} from "./Interface";
import {Serializer} from "./Serializer";

export abstract class TransportAbstract extends EventEmitter {
    public readonly serializer: Serializer;
    public readonly options: ConnectionOptions;

    constructor(options: ConnectionOptions) {
        super();

        this.options = options;
        this.serializer = new Serializer(options);
    }

    public send(data: object) {
        return this.enqueue(Buffer.from(JSON.stringify(data), "utf-8"));
    }

    public close() {
        this.destroy();
    }

    protected async enqueue(data: Buffer) {
        try {
            for (const chunk of await this.serializer.serialize(data)) {
                this.write(chunk);
            }
        } catch (error) {
            this.emit("error", error);
        }
    }

    protected abstract write(data: Buffer): void;

    protected abstract destroy(): void;
}

export type TransportCtor<T extends TransportAbstract = TransportAbstract> = new(options: ConnectionOptions) => T;
