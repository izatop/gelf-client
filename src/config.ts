import * as assert from "assert";
import {parse} from "url";
import {promisify} from "util";
import {deflate, ZlibOptions} from "zlib";
import {ConnectionOptions} from "./Interface";
import {TCPTransport, UDPTransport} from "./Transport";
import {TransportCtor} from "./TransportAbstract";

// @see https://docs.graylog.org/en/3.1/pages/gelf.html#chunking
export const GELF_MAGIC_NO = [0x1e, 0x0f];

export const MAX_CHUNK_SIZE = 1400;
export const MIN_COMPRESS_SIZE = 1400;

// Strict checks
export const STRICT_CHECKS = process.env.NODE_ENV !== "production";

const getIntValue = (value: any, min: number, max: number, defaultValue: number) => {
    if (!value) {
        return defaultValue;
    }

    const intValue = +value;
    if (intValue !== intValue || intValue < min || intValue > max) {
        return defaultValue;
    }

    return Math.round(intValue);
};

// compress is a promised zlib.deflate function
export const compress: (data: Buffer, options?: ZlibOptions) => Promise<Buffer> = promisify(deflate);

// transports
export const Transport = {
    transports: new Map<string, TransportCtor>([
        ["tcp", TCPTransport],
        ["udp", UDPTransport],
    ]),
    add(proto: string, transport: TransportCtor) {
        this.transports.set(proto, transport);
    },
    get(proto: string): TransportCtor {
        assert(
            typeof proto === "string" && this.transports.has(proto),
            `Transport for protocol ${proto} doesn't exists`,
        );

        return this.transports.get(proto)!;
    },
};

export const Timestamp = {
    get now() {
        return Math.round(Date.now() / 1000);
    },
};

export const parseConnectionString = (dsn: string): ConnectionOptions => {
    const config = parse(dsn);
    const options = new URLSearchParams(config.query as string || "");
    const alwaysStrictChecks = options.has("strict") || STRICT_CHECKS;

    assert(config.hostname, `Empty hostname`);
    assert(config.protocol, `Empty protocol`);

    // Ports 1024-49151 are the User Ports and are the ones to use for your own protocols.
    const port = getIntValue(config.port, 1024, 49151, 12201);
    const host = config.hostname!;
    const protocol = config.protocol!.substr(0, config.protocol!.length - 1);

    const maxChunkSize = getIntValue(
        options.get("maxChunkSize"),
        MAX_CHUNK_SIZE,
        Number.MAX_SAFE_INTEGER,
        MAX_CHUNK_SIZE,
    );

    const minCompressSize = getIntValue(
        options.get("minCompressSize"),
        MIN_COMPRESS_SIZE,
        Number.MAX_SAFE_INTEGER,
        MIN_COMPRESS_SIZE,
    );

    return {
        port,
        host,
        protocol,
        minCompressSize,
        maxChunkSize,
        compress: options.has("compress"),
        strictChecks: options.has("strict") || STRICT_CHECKS,
    };
};
