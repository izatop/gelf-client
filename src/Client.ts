import {parseConnectionString, Timestamp, Transport} from "./config";
import {ConnectionOptions, IEnvelope, IMessage, IMessageRest, MessageDefaults} from "./Interface";
import {Level} from "./Level";
import {TransportAbstract, TransportCtor} from "./TransportAbstract";

export class Client {
    public readonly version = "1.1";
    public readonly options: ConnectionOptions;
    public readonly transport: TransportAbstract;
    public readonly defaults: MessageDefaults = {};

    private constructor(transport: TransportAbstract, options: ConnectionOptions, defaults: MessageDefaults = {}) {
        this.transport = transport;
        this.options = options;
        this.defaults = defaults;
    }

    public static factory(dsn: string, defaults: MessageDefaults = {}): Client {
        const options = parseConnectionString(dsn);
        const transport = Transport.get(options.protocol);
        return new this(new transport(options), options, defaults);
    }

    public get isStrict() {
        return this.options.strictChecks;
    }

    public clone(defaults: MessageDefaults = {}) {
        return new Client(this.transport, this.options, {...this.defaults, ...defaults});
    }

    public send(data: IMessage & IMessageRest) {
        const {level, message, description, file, line, app, ...rest} = data;
        if (this.isStrict) {
            this.strictChecks(Object.keys(rest));
        }

        const timestamp = Timestamp.now;
        const envelope: IEnvelope = {
            host: app,
            level: level || Level.INFO,
            version: this.version,
            short_message: message,
            full_message: description,
            timestamp,
            file,
            line,
        };

        for (const [key, value] of Object.entries(rest)) {
            envelope[`_${key}`] = value;
        }

        return this.transport.send(envelope);
    }

    public emergency(data: Exclude<IMessage, "level"> & IMessageRest) {
        return this.send({...data, level: Level.EMERGENCY});
    }

    public alert(data: Exclude<IMessage, "level"> & IMessageRest) {
        return this.send({...data, level: Level.ALERT});
    }

    public critical(data: Exclude<IMessage, "level"> & IMessageRest) {
        return this.send({...data, level: Level.CRITICAL});
    }

    public error(data: Exclude<IMessage, "level"> & IMessageRest) {
        return this.send({...data, level: Level.ERROR});
    }

    public warning(data: Exclude<IMessage, "level"> & IMessageRest) {
        return this.send({...data, level: Level.WARNING});
    }

    public notice(data: Exclude<IMessage, "level"> & IMessageRest) {
        return this.send({...data, level: Level.NOTICE});
    }

    public info(data: Exclude<IMessage, "level"> & IMessageRest) {
        this.send({...data, level: Level.INFO});
    }

    public debug(data: Exclude<IMessage, "level"> & IMessageRest) {
        return this.send({...data, level: Level.DEBUG});
    }

    public close() {
        this.transport.close();
    }

    private strictChecks(keys: string[]) {
        const regex = /^([a-zA-Z0-9_])+$/;
        if (keys.some((key) => !regex.test(key))) {
            throw new Error(`Custom fields strict checks failed: ${keys.join(", ")}`);
        }
    }
}
