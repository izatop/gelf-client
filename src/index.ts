import {Client} from "./Client";

export {Client};
export {Level} from "./Level";
export * from "./Interface";
export * from "./TransportAbstract";

export default Client;
