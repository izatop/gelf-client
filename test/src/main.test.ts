import {Client, Level} from "../../src";
import {Timestamp, Transport} from "../../src/config";
import {TestTransport} from "./TestTransport";

const dsn = "test://localhost:123";
const clientDefaults = {foo: 1};
const time = 1574345698437;
const timestamp = Math.round(time / 1000);

Transport.add("test", TestTransport);
Object.defineProperty(Timestamp, "now", {get: () => timestamp});

test("Add transport", () => {
    expect(Timestamp.now).toBe(timestamp);
    expect(Transport.get("test")).toBe(TestTransport);
});

test("Client factory", async () => {
    const client = Client.factory(dsn, clientDefaults);
    expect(client.version).toBe("1.1");
    expect(client.transport).toBeInstanceOf(TestTransport);
    expect(client.defaults).toEqual(clientDefaults);
});

test("Strict checks", async () => {
    const client = Client.factory(dsn, clientDefaults);
    expect(() => client.send({"message": "test", "@wrong": 123})).toThrow();
});

test("Client send", async () => {
    const client = Client.factory(dsn, clientDefaults);
    await client.send({message: "test"});
    const transport = client.transport as TestTransport;
    const [written] = transport.written;
    expect(written).toBeInstanceOf(Buffer);
    expect(written.length).toBeGreaterThan(0);
    expect(JSON.parse(written.toString("utf-8"))).toEqual({
        timestamp,
        level: Level.INFO,
        short_message: "test",
        version: "1.1",
    });
});

test("Test chunks", async () => {
    const client = Client.factory(dsn, clientDefaults);
    await client.send({message: "chunk test", description: "foo".repeat(1100)});
    const transport = client.transport as TestTransport;
    const written = transport.written!;
    const expectChunksCount = 3;
    const expectMessageSize = 3397;
    expect(written.length).toBe(expectChunksCount);

    const chunks = new Array(written.length).fill(null);
    for (let index = 0; index < chunks.length; index++) {
        const chunk = written[index];
        const sequence = chunk.readInt8(10);
        const chunksCount = chunk.readInt8(11);
        chunks[sequence] = chunk.slice(12, chunk.length);
        expect(sequence).toBe(index);
        expect(chunksCount).toBe(expectChunksCount);
    }

    const message = Buffer.concat(chunks)
        .toString("utf-8");

    expect(message.length).toBe(expectMessageSize);
    expect(JSON.parse(message)).toMatchSnapshot();
});
