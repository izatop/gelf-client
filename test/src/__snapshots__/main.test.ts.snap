// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test chunks 1`] = `
Object {
  "full_message": "foofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoo",
  "level": 6,
  "short_message": "chunk test",
  "timestamp": 1574345698,
  "version": "1.1",
}
`;
